utils/serde/src/main/java/io/apicurio/registry/utils/serde/JsonSchemaKafkaSerializer.java

```diff
@@ -0,0 +1,192 @@
+            return new byte[0];
+        }
+
+        // Now serialize the data
+        try {
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
+            JsonGenerator generator = mapper.getFactory().createGenerator(baos);
+            if (validationEnabled) {
+                String artifactId = getArtifactId(topic, data);
+                Integer version = getArtifactVersion(artifactId, topic, data);
+                addSchemaHeaders(headers, artifactId, version);
+
+                String schema = loadSchema(artifactId, version);
+                SchemaValidator schemaValidator = api.loadSchema(new StringSchemaSource(schema));
+                // TODO cache the SchemaValidator - keyed on artifactId+version
+                generator = api.decorateJsonGenerator(schemaValidator, generator);
+            }
+            addTypeHeaders(headers, data);
+
+            mapper.writeValue(generator, data);
+            
+            return baos.toByteArray();
+        } catch (IOException e) {
+            throw new SerializationException(e);
+        }
+    }
+
+    /**
+     * Figure out the artifact ID from the topic name and data.
+     * @param topic
+     * @param data
+     */
+    private String getArtifactId(String topic, T data) {
+        // TODO support other options - other impls use a strategy for this
+        return topic;
+    }
+
+    /**
+     * Figure out which version of the artifact to use, specifically.  If this returns
+     * null then the latest version will be used.
+     * @param artifactId
+     * @param topic
+     * @param data
+     */
+    private Integer getArtifactVersion(String artifactId, String topic, T data) {
+        // TODO how can we know what artifact version we need?  passed in via config?
+        return null;
+    }
+
+    /**
+     * Adds appropriate information to the Headers so that the deserializer can function properly.
+     * @param headers
+     * @param artifactId
+     * @param version
+     */
+    private void addSchemaHeaders(Headers headers, String artifactId, Integer version) {
+        headers.add(JsonSchemaSerDeConstants.HEADER_ARTIFACT_ID, artifactId.getBytes(StandardCharsets.UTF_8));
+        if (version != null) {
+            ByteBuffer buff = ByteBuffer.allocate(4);
+            buff.putInt(version.intValue());
+            headers.add(JsonSchemaSerDeConstants.HEADER_VERSION, buff.array());
+        }
+    }
+
+    /**
+     * Adds appropriate information to the Headers so that the deserializer can function properly.
+     * @param headers
+     * @param data
+     */
+    private void addTypeHeaders(Headers headers, T data) {
+        headers.add(JsonSchemaSerDeConstants.HEADER_MSG_TYPE, data.getClass().getName().getBytes(StandardCharsets.UTF_8));
+    }
+
+    /**
+     * Gets the schema from the registry.
+     * @param artifactId
+     * @param version
+     */
+    private String loadSchema(String artifactId, Integer version) {
```

Comment:
This is a dup from deserializer ... perhaps create a common class for both.
And use SchemaCache to cache the results ...